corporate/lib/stripe.py

```diff
@@ -478,6 +508,42 @@ def invoice_plan(plan: CustomerPlan, event_time: datetime) -> None:
         plan.save(update_fields=['invoicing_status', 'invoiced_through'])
         licenses_base = ledger_entry.licenses
 
+    if plan.status == CustomerPlan.ENDED:
+        new_plan = get_current_plan_by_realm(plan.customer.realm)
+
+        if new_plan is not None and new_plan.invoicing_status == CustomerPlan.INITIAL_INVOICE_TO_BE_SENT:
+            ledger_entry = LicenseLedger.objects.filter(plan=new_plan, event_time__lte=event_time,
+                                                        is_renewal=True).order_by("id").first()
+
+            if new_plan.fixed_price is not None:
+                price_args = {'amount': new_plan.fixed_price}
+            else:
+                assert(new_plan.price_per_license is not None)  # needed for mypy
+                price_args = {'unit_amount': new_plan.price_per_license,
+                              'quantity': ledger_entry.licenses}
+
+            if new_plan.billing_schedule == CustomerPlan.ANNUAL:
+                period_end = add_months(new_plan.billing_cycle_anchor, 12)
+            else:  # nocoverage
+                period_end = add_months(new_plan.billing_cycle_anchor, 1)
+
+            idempotency_key = f'ledger_entry:{ledger_entry.id}'
+            if settings.TEST_SUITE:
+                idempotency_key = None
+            stripe.InvoiceItem.create(
+                currency='usd',
+                customer=new_plan.customer.stripe_customer_id,
+                description='Zulip Standard - upgrade to annual plan',
+                discountable=False,
+                period = {'start': datetime_to_timestamp(new_plan.billing_cycle_anchor),
+                          'end': datetime_to_timestamp(period_end)},
+                idempotency_key=idempotency_key,
+                **price_args)
```

Comment:
this idempotency_key logic seems like should be extracted into some kind of helper, since we have an analogical block higher up in this function?

Reply:
Done